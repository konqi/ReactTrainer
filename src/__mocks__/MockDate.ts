export class MockDate extends Date {
  constructor(datetime?: number) {
    super(datetime || 3141592653589)
    return this
  }

  static now() {
    return 3141592653589
  }
}
