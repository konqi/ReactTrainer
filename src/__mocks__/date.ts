const originalDate = global.Date

export const freezeTime = (clockFrozenAt: number = 3141592653589): Date =>
  //@ts-ignore
  class extends originalDate {
    constructor(datetime?: number | string | Date) {
      super((typeof datetime !== undefined && datetime) || clockFrozenAt)
      return this
    }

    static now() {
      return clockFrozenAt
    }
  }

export const restoreCausality = () => originalDate
