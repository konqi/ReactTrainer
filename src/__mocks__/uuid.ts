export const v1 = () => "1";
export const v4 = () => "4";
