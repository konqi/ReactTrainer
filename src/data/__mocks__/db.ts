export const deleteMock = jest.fn()
export const docMock = jest.fn()

export const db = {
  batch: () => ({
    delete: deleteMock,
  }),
  collection: () => ({
    doc: docMock,
  }),
}
