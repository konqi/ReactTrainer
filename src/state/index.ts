import {TraineeState, TraineeFSAs} from './trainees/types'
import {initialTraineeState} from './trainees'
import {UiState, UiFSAs} from './ui/types'
import {TrainingFSAs} from './trainings/types'
import {initialUiState} from './ui/uiReducer'
import {createStore, applyMiddleware} from 'redux'
import {rootReducer} from './rootReducer'
import createSagaMiddleware from 'redux-saga'
import {rootSaga} from './sagas/rootSaga'

export interface ApplicationState {
  trainees: TraineeState
  ui: UiState
}

export const initialApplicationState: ApplicationState = {
  trainees: initialTraineeState,
  ui: initialUiState,
}

export type Action = TraineeFSAs | TrainingFSAs | UiFSAs

const sagaMiddleware = createSagaMiddleware()

export const store = createStore(
  rootReducer,
  initialApplicationState,
  applyMiddleware(sagaMiddleware)
)

sagaMiddleware.run(rootSaga)
