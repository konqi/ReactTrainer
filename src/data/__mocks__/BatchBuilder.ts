import {Collection} from '../../types/Collection'
import {db} from '../db'

const batchDeleteMock = jest.fn()
const batchCommitMock = jest.fn()

const BatchBuilderMock = {
  delete: batchDeleteMock,
  execute: batchCommitMock,
}

export class BatchBuilder {
  constructor() {
    return BatchBuilderMock
  }
}
