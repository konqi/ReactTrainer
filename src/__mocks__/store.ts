import {createStore} from 'redux'
import {ApplicationState} from '../state'
import {rootReducer} from '../state/rootReducer'
import {Trainee} from '../types/Trainee'
import {Session} from '../types/Session'

export const createMockStore = (initialApplicationState: ApplicationState) =>
  createStore(rootReducer, initialApplicationState)

export class TraineeBuilder {
  trainee: Trainee
  constructor() {
    this.trainee = {
      id: 'traineeId',
      name: 'Dummy Trainee',
      price: 15,
    } as Trainee
  }

  withId(id: string) {
    this.trainee.id = id
    return this
  }

  withName(name: string) {
    this.trainee.name = name
    return this
  }

  withPrice(price: number) {
    this.trainee.price = price
    return this
  }

  withSessionRef(sessionRef: string[]) {
    this.trainee.sessionsRef = sessionRef
    return this
  }

  build() {
    return this.trainee
  }
}

export class SessionBuilder {
  session: Session

  constructor(traineeId: string) {
    this.session = {
      id: 'sessionId',
      datetime: new Date(),
      payedAmount: 15,
      traineeRef: traineeId,
    }
  }

  withId(id: string) {
    this.session.id = id
    return this
  }

  withDatetime(datetime: Date) {
    this.session.datetime = datetime
    return this
  }

  withDescription(description: string) {
    this.session.description = description
    return this
  }

  withPayedAmount(payedAmount: number) {
    this.session.payedAmount = payedAmount
    return this
  }

  withPrice(price: number) {
    this.session.price = price
    return this
  }

  build() {
    return this.session
  }
}
