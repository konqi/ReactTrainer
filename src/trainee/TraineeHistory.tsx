import React from 'react'
import {
  Container,
  Paper,
  createStyles,
  Theme,
  makeStyles,
  Typography,
  Table,
  TableRow,
  TableCell,
  TableBody,
  TableHead,
} from '@material-ui/core'
import {ApplicationState} from '../state'
import {useSelector} from 'react-redux'
import {Trainee} from '../types/Trainee'
import get from 'lodash/get'
import {Session} from '../types/Session'
import format from 'date-fns/format'
import {red, green} from '@material-ui/core/colors'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      margin: theme.spacing(1, 0),
    },
    negativeBalance: {
      color: red[400],
    },
    positiveBalance: {
      color: green[400],
    },
    pullRight: {
      position: 'absolute',
      right: 0,
    },
    payedColumn: {
      position: 'relative',
    },
  })
)

interface ExternalProps {
  traineeId: string
}
export const TraineeHistory: React.FC<ExternalProps> = ({
  traineeId,
}: ExternalProps) => {
  const classes = useStyles()
  const trainee = useSelector<ApplicationState, Trainee>(
    state => state.trainees[traineeId]
  )
  const sessions = useSelector<ApplicationState, Session[]>(state => {
    const sessionRefs = get(trainee, 'sessionsRef')
    if (sessionRefs !== undefined && sessionRefs.length > 0) {
      return sessionRefs.map(sessionRef => state.sessions[sessionRef])
    } else {
      return []
    }
  })

  const balanceClass = (balance: number) => {
    if (balance > 0) {
      return classes.positiveBalance
    } else if (balance < 0) {
      return classes.negativeBalance
    }
  }

  return (
    <Paper className={classes.root}>
      <Container>
        <Typography variant="h5">Historie</Typography>
        <Table>
          <TableHead>
            <TableRow>
              <TableCell size="small">Datum</TableCell>
              <TableCell size="small">Preis</TableCell>
              <TableCell size="medium" className={classes.payedColumn}>
                Gezahlt (Differenz)
              </TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            {sessions.map(session => {
              const price = session.price || trainee.price
              const difference = session.payedAmount - price
              return (
                <TableRow key={session.id}>
                  <TableCell>{format(session.datetime, 'D.M.YY')}</TableCell>
                  <TableCell>{price}</TableCell>
                  <TableCell className={classes.payedColumn}>
                    {session.payedAmount}
                    {difference !== 0 && (
                      <span
                        className={`${balanceClass(difference)} ${
                          classes.pullRight
                        }`}
                      >
                        ({difference})
                      </span>
                    )}
                  </TableCell>
                </TableRow>
              )
            })}
          </TableBody>
        </Table>
      </Container>
    </Paper>
  )
}
