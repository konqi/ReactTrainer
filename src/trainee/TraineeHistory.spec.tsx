import React from 'react'
import {render, cleanup} from '@testing-library/react'
import {TraineeHistory, TraineeHistoryConnected} from './TraineeHistory'
import {
  TraineeBuilder,
  SessionBuilder,
  createMockStore,
} from '../__mocks__/store'
import {initialApplicationState} from '../state'
import {Provider} from 'react-redux'
import {restoreCausality, freezeTime} from '../__mocks__/date'

describe('snapshot tests', () => {
  beforeAll(() => {
    // @ts-ignore
    global.Date = freezeTime(3141592653589)
  })

  afterAll(() => {
    global.Date = restoreCausality()
  })
  afterEach(cleanup)
  it('renders table with sessions', () => {
    const trainee = new TraineeBuilder().withPrice(10).build()
    const session1 = new SessionBuilder(trainee.id)
      .withId('Session1')
      .withPayedAmount(5)
      .build()
    const session2 = new SessionBuilder(trainee.id)
      .withId('Session2')
      .withPayedAmount(10)
      .build()
    const session3 = new SessionBuilder(trainee.id)
      .withId('Session3')
      .withPayedAmount(15)
      .build()

    const {baseElement} = render(
      <TraineeHistory
        trainee={trainee}
        sessions={[session1, session2, session3]}
      />
    )

    expect(baseElement).toMatchSnapshot()
  })
})

describe('integration tests', () => {
  beforeAll(() => {
    // @ts-ignore
    global.Date = freezeTime(3141592653589)
  })

  afterAll(() => {
    global.Date = restoreCausality()
  })
  afterEach(cleanup)
  it('renders table with session data from store', () => {
    const trainee = new TraineeBuilder().withPrice(10).build()
    const session1 = new SessionBuilder(trainee.id)
      .withId('Session1')
      .withPayedAmount(5)
      .build()
    const session2 = new SessionBuilder(trainee.id)
      .withId('Session2')
      .withPayedAmount(10)
      .build()
    const session3 = new SessionBuilder(trainee.id)
      .withId('Session3')
      .withPayedAmount(15)
      .build()
    trainee.sessionsRef = [session1.id, session2.id, session3.id]

    const store = createMockStore({
      ...initialApplicationState,
      trainees: {
        [trainee.id]: trainee,
      },
      sessions: {
        [session1.id]: session1,
        [session2.id]: session2,
        [session3.id]: session3,
      },
    })

    const {baseElement} = render(
      <Provider store={store}>
        <TraineeHistoryConnected traineeId={trainee.id} />
      </Provider>
    )

    expect(baseElement).toMatchSnapshot()
  })
})
