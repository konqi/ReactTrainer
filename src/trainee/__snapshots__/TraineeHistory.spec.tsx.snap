// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration tests renders table with session data from store 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation1 makeStyles-root-89 MuiPaper-rounded"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5"
        >
          Historie
        </h5>
        <table
          class="MuiTable-root"
        >
          <thead
            class="MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
                scope="col"
              >
                Datum
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
                scope="col"
              >
                Preis
              </th>
              <th
                class="MuiTableCell-root makeStyles-payedColumn-93 MuiTableCell-head"
                scope="col"
              >
                Gezahlt (Differenz)
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root"
          >
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                21.7.69
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                10
              </td>
              <td
                class="MuiTableCell-root makeStyles-payedColumn-93 MuiTableCell-body"
              >
                5
                <span
                  class="makeStyles-negativeBalance-90 makeStyles-pullRight-92"
                >
                  (
                  -5
                  )
                </span>
              </td>
            </tr>
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                21.7.69
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                10
              </td>
              <td
                class="MuiTableCell-root makeStyles-payedColumn-93 MuiTableCell-body"
              >
                10
              </td>
            </tr>
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                21.7.69
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                10
              </td>
              <td
                class="MuiTableCell-root makeStyles-payedColumn-93 MuiTableCell-body"
              >
                15
                <span
                  class="makeStyles-positiveBalance-91 makeStyles-pullRight-92"
                >
                  (
                  5
                  )
                </span>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</body>
`;

exports[`snapshot tests renders table with sessions 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation1 makeStyles-root-1 MuiPaper-rounded"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5"
        >
          Historie
        </h5>
        <table
          class="MuiTable-root"
        >
          <thead
            class="MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
                scope="col"
              >
                Datum
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
                scope="col"
              >
                Preis
              </th>
              <th
                class="MuiTableCell-root makeStyles-payedColumn-5 MuiTableCell-head"
                scope="col"
              >
                Gezahlt (Differenz)
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root"
          >
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                21.7.69
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                10
              </td>
              <td
                class="MuiTableCell-root makeStyles-payedColumn-5 MuiTableCell-body"
              >
                5
                <span
                  class="makeStyles-negativeBalance-2 makeStyles-pullRight-4"
                >
                  (
                  -5
                  )
                </span>
              </td>
            </tr>
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                21.7.69
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                10
              </td>
              <td
                class="MuiTableCell-root makeStyles-payedColumn-5 MuiTableCell-body"
              >
                10
              </td>
            </tr>
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                21.7.69
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
              >
                10
              </td>
              <td
                class="MuiTableCell-root makeStyles-payedColumn-5 MuiTableCell-body"
              >
                15
                <span
                  class="makeStyles-positiveBalance-3 makeStyles-pullRight-4"
                >
                  (
                  5
                  )
                </span>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</body>
`;
